src/components/tl/mlx5/mcast/tl_mlx5_mcast_helper.h

```diff
@@ -0,0 +1,322 @@
+            ucc_list_del(&pp->super);
+            ucc_tl_mlx5_mcast_process_packet(comm, req, pp);
+            num_left--;
+        } else if (pp->psn < comm->last_acked){
+            ucc_list_del(&pp->super);
+            ucc_list_add_tail(&comm->bpool, &pp->super);
+        }
+
+        (*pending_q_size)++;
+    };
+
+    wc = ucc_malloc(sizeof(struct ibv_wc) * POLL_PACKED, "WC");
+    if (!wc) {
+        tl_error(comm->lib, "ucc_malloc failed");
+        return -1;
+    }
+
+    while (num_left > 0)
+    {
+        memset(wc, 0, sizeof(struct ibv_wc) * POLL_PACKED);
+        num_comp = ibv_poll_cq(comm->rcq, POLL_PACKED, wc);
+
+        if (num_comp < 0) {
+            tl_error(comm->lib, "recv queue poll completion failed %d", num_comp);
+            ucc_free(wc);
+            return -1;
+        } else if (num_comp == 0) {
+            break;
+        }
+
+        real_num_comp = num_comp;
+
+        for (i = 0; i < real_num_comp; i++) {
+            if (IBV_WC_SUCCESS != wc[i].status) {
+                tl_error(comm->lib, "mcast_recv: %s err pending_recv %d wr_id %ld"
+                         " num_comp %d byte_len %d",
+                         ibv_wc_status_str(wc[i].status), comm->pending_recv,
+                         wc[i].wr_id, num_comp, wc[i].byte_len);
+                ucc_free(wc);
+                return -1;
+            }
+
+            id         = wc[i].wr_id;
+            pp         = (struct pp_packet*) (id);
+            pp->length = wc[i].byte_len - GRH_LENGTH;
+            pp->psn    = ntohl(wc[i].imm_data);
+
+            tl_trace(comm->lib, "completion: psn %d, length %d, already_received %d, "
+                                " psn in req %d, req_start %d, req_num packets"
+                                " %d, to_send %d, to_recv %d, num_left %d",
+                                pp->psn, pp->length, PSN_RECEIVED(pp->psn,
+                                comm) > 0, PSN_IS_IN_RANGE(pp->psn, req,
+                                comm), req->start_psn, req->num_packets,
+                                req->to_send, req->to_recv, num_left);
+
+            if (PSN_RECEIVED(pp->psn, comm) || pp->psn < comm->last_acked) {
+                /* This psn was already received */
+                ucc_assert(pp->context == 0);
+                ucc_list_add_tail(&comm->bpool, &pp->super);
+            } else {
+                if (num_left > 0 && PSN_IS_IN_RANGE(pp->psn, req, comm)) {
+                    __builtin_prefetch(req->ptr + PSN_TO_RECV_OFFSET(pp->psn, req, comm));
+                    __builtin_prefetch((void*) pp->buf);
+                    ucc_tl_mlx5_mcast_process_packet(comm, req, pp);
+                    num_left--;
+                } else {
+                    ucc_list_add_tail(&comm->pending_q, &pp->super);
+                }
+            }
+        }
+
+        comm->pending_recv -= num_comp;
+        ucc_tl_mlx5_mcast_post_recv_buffers(comm);
+    }
+
+    ucc_free(wc);
+    return num_left;
```

Comment:
description says "this function return the number of mcast recv packets"
seems like it returns something different